sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala

```diff
@@ -239,6 +239,15 @@ object HiveThriftServer2 extends Logging {
       executionList(id).state = ExecutionState.COMPILED
     }
 
+    def onStatementCanceled(id: String): Unit = {
+      synchronized {
+        executionList(id).finishTimestamp = System.currentTimeMillis
+        executionList(id).state = ExecutionState.CANCELED
+        totalRunning -= 1
```

Comment:
It can happen that `cancel()` will race with query finishing, and then both `onStatementCanceled()` and `onStatementFinish()` will do `totalRunning -= 1`, resulting in negative count.

Instead of keeping a counter, for sessions and queries, it is better to do something like:
```
    def getOnlineSessionNum: Int = synchronized {
      sessionList.count(_._2.finishTimestamp == 0)
    }

    def isExecutionActive(execInfo: ExecutionInfo): Boolean = {
      !(execInfo.state == ExecutionState.FAILED ||
        execInfo.state == ExecutionState.CANCELED ||
        execInfo.closeTimestamp != 0)
    }

    /**
     * When an error or a cancellation occurs, we set the finishTimestamp of the statement.
     * Therefore, when we count the number of running statements, we need to exclude errors and
     * cancellations and count all statements that have not been closed so far.
     */
    def getTotalRunning: Int = synchronized {
      executionList.count {
        case (_, v) => isExecutionActive(v)
      }
    }
```
(Note that we should also consider statements that are CLOSED instead of FINISHED as not running anymore, because a FINISHED statement is still fetching results.)
(Same for sessions counters - though quite an obscure, there can be a race between a session timeout and a client closing session)

Reply:
@juliuszsompolski  if we call close() before FINISH, onOperationClose() won't do 
`totalRunning -= 1`, also a bug.

Reply:
@juliuszsompolski Can help to look at logic again

Reply:
@AngersZhuuuu I think it's easy to get something wrong with the `totalRunning` counter. Even if we covered all the cases now, it's hard to reason about all the possible paths, and easy to break. I would suggest to count it on the fly based on status, like I proposed before.
```
    def getOnlineSessionNum: Int = synchronized {
      sessionList.count(_._2.finishTimestamp == 0)
    }

    def isExecutionActive(execInfo: ExecutionInfo): Boolean = {
      !(execInfo.state == ExecutionState.FAILED ||
        execInfo.state == ExecutionState.CANCELED ||
        execInfo.closeTimestamp != 0)
    }

    /**
     * When an error or a cancellation occurs, we set the finishTimestamp of the statement.
     * Therefore, when we count the number of running statements, we need to exclude errors and
     * cancellations and count all statements that have not been closed so far.
     */
    def getTotalRunning: Int = synchronized {
      executionList.count {
        case (_, v) => isExecutionActive(v)
      }
    }
```

Reply:
`def isExecutionActive(execInfo: ExecutionInfo): Boolean = { `
 `     !(execInfo.state == ExecutionState.FAILED ||`
  `      execInfo.state == ExecutionState.CANCELED ||`
 `       execInfo.state == ExecutionState.CLOSED)`
  ` }`

make  **isExecutionActive()** like this? to keep same code style

Reply:
It can cover all situation, very good ideal